python/lsst/daf/butler/core/queries/builder.py

```diff
@@ -0,0 +1,400 @@
+        automatically call it if the `DimensionElement` has been identified as
+        one that must be included.
+
+        Parameters
+        ----------
+        element : `DimensionElement`
+            Element for which a table should be added.  The element must be
+            associated with a database table (see `DimensionElement.hasTable`).
+        """
+        assert element not in self._elements, "Element already included in query."
+        assert element.hasTable(), "Cannot join element with no table."
+        table = self._dimensionStorage[element].getElementTable(self.summary.dataId)
+        if element in self.summary.spatial:
+            self.joinToCommonSkyPix(element)
+            self._columns.regions[element] = table.columns[REGION_FIELD_SPEC.name]
+        joinDimensions = list(element.graph.required)
+        joinDimensions.extend(element.implied)
+        joinOn = self._startJoin(table, joinDimensions, element.RecordClass.__slots__)
+        if element in self.summary.temporal:
+            intervalInTable = Timespan(
+                begin=table.columns[TIMESPAN_FIELD_SPECS.begin.name],
+                end=table.columns[TIMESPAN_FIELD_SPECS.end.name],
+            )
+            for intervalInQuery in self._columns.timespans.values():
+                joinOn.append(intervalInQuery.overlaps(intervalInTable, ops=sqlalchemy.sql))
+            self._columns.timespans[element] = intervalInTable
+        self._finishJoin(table, joinOn)
+        self._elements[element] = table
+
+    def joinToCommonSkyPix(self, element: DimensionElement):
+        """Add the table relating a spatial `DimensionElement` to the
+        universe's `~DimensionUniverse.commonSkyPix` dimension to the query.
+
+        External calls to this method should rarely be necessary; `finish` will
+        automatically call it if the `DimensionElement` has been identified as
+        one that must be included.
+
+        Parameters
+        ----------
+        element : `DimensionElement`
+            Element for which the relationship should be added.  The element
+            must be associated with a database table (see
+            `DimensionElement.hasTable`) and must be
+            `~Dimensionelement.spatial`.
+        """
+        if element is self.summary.universe.commonSkyPix:
+            return
+        assert element in self.summary.spatial,\
+            f"{element}'s spatial information is not relevant for this query."
+        table = self._dimensionStorage[element].getCommonSkyPixOverlapTable(self.summary.dataId)
+        assert table is not None, f"No relationship found from the common skypix to {element}."
+        dimensions = NamedValueSet(element.graph.required)
+        dimensions.add(self.summary.universe.commonSkyPix)
+        self.joinTable(table, dimensions)
+
+    def joinDataset(self, datasetType: DatasetType, collections: CollectionsExpression, *,
+                    isResult: bool = True, addRank: bool = False):
```

Comment:
No, actually - this is an unfortunate difference between `typing.Optional` (which means "may be `None`") and the less-formal usage of `<type>, optional` in docstrings as specified by the dev guide (which means "need not be passed by caller").


This might be something @jonathansick might be interested in thinking about, if he hasn't already, though I don't think there's an easy answer: both meanings are useful and precisely defined, and are arguably each the most useful meaning for their own slightly-different contexts.